### ISSUE TITLE
Short Prefixes

### ISSUE THREAD

**Issue by guy-who-googles** [twpayne/chezmoi#1313]
## Is your feature request related to a problem? Please describe.

Would adding short, predetermined prefixes to represent attributes be of benefit? They'd be slightly longer than abbreviations, but shorter than the full name.

## Describe the solution you'd like

| Current prefix | Short prefix suggestion | Reasoning |
| ---            | ---                     | ---       |
| after          | after | easier to understand workflow without change |
| before         | before | easier to understand workflow without change |
| create         | cre | |
| dot            | dot | short enough as is |
| empty          | emp | |
| encrypted      | enc | |
| exact          | exa | |
| executable     | exe | |
| literal        | lit | |
| modify         | mod | |
| once           | once | easier to understand workflow without change |
| private        | pri | |
| run            | run | short enough as is |
| symlink        | sym | |

To invoke, do:
| Command | Name|
| ---     | ---      |
| chezmoi chattr exa,pri,dot ~/.dir | exa_pri_dot_dir |
| chezmoi chattr exact,private,dot ~/.dir | exact_private_dot_dir |

## Describe alternatives you've considered

* While custom structures would be nice, being cross-platform and simple is of higher value (as other issues have already discussed)
* Rely on other software or create another repo for syncing non-dotfiles. This adds complexity to cross machine setup, but makes nested directories easier to work with
* Using abbreviations as prefixes, but that would be difficult to remember, even after extensive usage. For example, `epd_dir` = `exact_private_dot_dir`.
* Use `chezmoi edit` - I find rapid context switching more difficult because using my editor's file finder opens the target file. This requires an extra step of reflecting those changes back to the source state through merging or re-adding the file.
* Chezmoi is brilliant already, just deal with the long prefixes :)

## Additional context

It may help to clarify that, in addition to using chezmoi as a dotfile manager, I currently use it to manage all cross-machine syncing. This includes
* notes - could easily be moved to a different repo/sync method. Chezmoi permission handling is nice though.
* [navi](https://github.com/denisidoro/navi) cheats - templating is useful. When you don't have all the same tools, like copying to clipboard, using template variables to accomplish the same functionality with slightly different tools in the same location makes things much more manageable.

Using chezmoi in this way introduces two new challenges:
1. Directory depths can be several levels much more easily. As a dotfile manager, I don't imagine many users have more than 3 levels (private_config/tool/tool.yaml). With notes and cheats, I get to 4 or 5 pretty quickly.
1. Working within your repo more frequently and searching for files. Configurations tend to lean towards set and forget, while note-taking tends to be more continuous.

While this isn't chezmoi's stated purpose, even without this suggestion, it easily could be. Shortening verbose file names makes it faster to understand and navigate their repo, at the cost of slightly obscuring filenames. For those working directly in their chezmoi repo often, this seems like a worthwhile trade-off. Even with fuzzy searching, everything takes a bit longer than you'd like because you're reading (in my case) exact_private on every directory level.

Several issues regarding long prefixes (usually about aesthetic distaste) have been closed. Since this is a real world use case, probably trivial to implement, and backwards-compatible, I thought I'd mention it. I don't want to belabor the point, so no worries if this still isn't of interest. Thank you for making such an elegant and usable solution to this problem :)

**Comment by twpayne** [comment]
Thank you for the very well thought through proposal - really appreciated.

Adding alternative spellings of prefixes would mean that a file with N attributes would have 2^N possible filenames in the source state. Of course, it's possible to keep track of whether the short or long attribute name is used for each attribute, but this starts to add complexity and ambiguity that I'd prefer to avoid at this point.

> It may help to clarify that, in addition to using chezmoi as a dotfile manager, I currently use it to manage all cross-machine syncing.

Using chezmoi for all cross-machine syncing is definitely pushing chezmoi beyond what it was initially designed for. Could you give a few more details about how you use it? There might be alternative solutions for your use case, e.g. storing your notes and navi cheats in source directory ignored by chezmoi (e.g. putting them in a subdirectory with a leading `.` or adding the notes/navi directory to `.chezmoiignore`) and using a chezmoi-managed symlink from your home directory to the ignored directory so that notes/navi finds the files at the right path.

Thank you again for your proposal - it's great to get well-considered contributions like this :)

**Comment by guy-who-googles** [comment]
> Thank you for the very well thought through proposal - really appreciated.

Of course, thank you for the well-designed software!

> Adding alternative spellings of prefixes would mean that a file with N attributes would have 2^N possible filenames in the source state. Of course, it's possible to keep track of whether the short or long attribute name is used for each attribute, but this starts to add complexity and ambiguity that I'd prefer to avoid at this point.

Pre-determining the short prefix naming scheme could limit that ambiguity. A file with N attributes would have these number of possible filenames:
1. 2*N, if a file could either have short or long attributes
1. 2, if there was a configuration option determining which scheme the user had to use for their entire repo. The default could be "long" for backwards-compatability. On every `chezmoi diff` there could be a check enforcing those schemes, giving a warning listing files not in compliance.

I didn't consider the additional complexity of handling mixed attribute sizes since I would've converted everything to short. You're right - this isn't as trivial to implement as I hoped. Barring a more popularized use case, I understand why you'd like to avoid it. I wrote out the suggestion in case that changes.

> Using chezmoi for all cross-machine syncing is definitely pushing chezmoi beyond what it was initially designed for. Could you give a few more details about how you use it? There might be alternative solutions for your use case, e.g. storing your notes and navi cheats in source directory ignored by chezmoi (e.g. putting them in a subdirectory with a leading . or adding the notes/navi directory to .chezmoiignore) and using a chezmoi-managed symlink from your home directory to the ignored directory so that notes/navi finds the files at the right path.

Yes, any help is greatly appreciated. Having chezmoi ignore my notes directory is actually a better solution than this proposal would've been. Thanks, didn't think of that :)<br>
For the navi folder, that gets a bit more complicated since I make extensive use of templating. On the bright side, I can specify the path to my navi top-level directory, so symlinking shouldn't be a concern.<br>
To abstract the tool away, let's use aliasing as an example. For further simplicity, I'll make this silly copy alias and using chezmoi templating to make it consistent across different machines. My .bash_aliases would look something like:
```
{{ if (eq .chezmoi.hostname "wayland_linux_at_home") }}
alias my_cp="cat hello_world.txt | wl-copy"
{{ else if (eq .chezmoi.hostname "x11_linux_at_work") }}
alias my_cp="cat hello_world.txt | xclip"
{{ else if (eq .chezmoi.hostname "mac_at_work") }}
alias my_cp="cat hello_world.txt | pbcopy"
{{ end }}
```
Navi let's me do more advanced versions of this. As far as chezmoi is concerned, the main difference is that I have many of these, organized into a deeper directory structure. Each directory has attributes:
* exact - to make chezmoi the single source of truth
* private - with write access, someone could have me execute an alias to an aribtrary command I thought I wrote

If I ignore my navi directory, chezmoi won't interpret these templates.

To ease the burden of context switching, I usually `grep` the folder from the top level for a keyword. Doing so gives me results such as `exact_private_dot_navi/exact_private_tools/exact_private_chezmoi.cheat`. Manageable, but not too fun to deal with.
